### PR TITLE
Bonnmotion improvements

### LINKED ISSUE #61: Use docker or user-definied BonnMotion executable

### PR DESCRIPTION
**Pull request by ebinf**
# Description of the PR

Added flag to specify custom BonnMotion executable (fixes #61).

Reworked all tests to use `assert`, which makes them more readable and consistent.

Refactored and prepared for more movement patterns.

# Review checklist

(based on the work of Karl E. Wiegers)

**Structure**

- [ ] Does the code completely and correctly implement the design?
- [ ] Is the code well-structured, consistent in style, and consistently formatted?
- [ ] Are there any uncalled or unneeded procedures or any unreachable code?
- [ ] Are there any leftover stubs or test routines in the code?
- [ ] Can any code be replaced by calls to external reusable components or library functions?
- [ ] Are there any blocks of repeated code that could be condensed into a single procedure?
- [ ] Are symbolics used rather than “magic number” constants or string constants?
- [ ] Are any modules excessively complex and should be restructured or split into multiple routines?

**Documentation**

- [ ] Is the code clearly and adequately documented with an easy-to-maintain commenting style?
- [ ] Are all comments consistent with the code?

**Variables**

- [ ] Are all variables properly defined with meaningful, consistent, and clear names?
- [ ] Are there any redundant or unused variables?

**Arithmetic Operations**

- [ ] Does the code avoid comparing floating-point numbers for equality?
- [ ] Are divisors tested for zero or noise?

**Loops and Branches**

- [ ] Are all loops, branches, and logic constructs complete and correct?
- [ ] Are all loops, branches, and logic constructs properly nested and avoid unnecessary nesting?
- [ ] Can any statements that are enclosed within loops be placed outside the loops?

**Defensive Programming**

- [ ] Are imported data and input arguments tested for validity and completeness?
- [ ] Are files checked for existence before attempting to access them?

### CODE REVIEW

File: internal/outputgenerators/bonnmotion.go [hunk 1]
```diff
@@ -114,36 +110,48 @@ func (b Bonnmotion) execute(command []string) error {
...
 }
 
-// Calls BonnMotion to generate the Random Waypoint data for a given NodeGroup inside an Experiment.
-func (b Bonnmotion) generateRandomWaypointNodeGroup(exp experiment.Experiment, nodeGroup experiment.NodeGroup) {
-	logger.Trace("Generating Random Waypoint movements")
+// Calls BonnMotion to generate the for a given NodeGroup inside an Experiment.
+func (b Bonnmotion) generateNodeGroup(exp experiment.Experiment, nodeGroup experiment.NodeGroup) error {
+	logger.Tracef("Generating \"%s\" movements", nodeGroup.MovementModel.String())
+	var movementModelParameters []string
+	switch movementModel := nodeGroup.MovementModel.(type) {
```
Comment (SA8R1NA) [inline internal/outputgenerators/bonnmotion.go:117]: please declare movementModel outside of the switch for a better readabillity

File: internal/outputgenerators/bonnmotion.go [hunk 2]
```diff
@@ -156,27 +164,43 @@ func (b Bonnmotion) Generate(exp experiment.Experiment) {
...
+	stepFilePath := filepath.Join(b.outputFolder, BonnMotionStepFile)
+	allowedToWriteStepFile := folderstructure.MayCreatePath(stepFilePath)
 	if !allowedToWriteStepFile {
 		logger.Error("Not allowed to write step file!")
 		return
 	}
+	b.stepFile, err = os.OpenFile(stepFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		logger.Error("Error opening step file:", err)
+		return
```
Comment (SA8R1NA) [inline internal/outputgenerators/bonnmotion.go:176]: please stay consistent with return and return nil

File: internal/outputgenerators/bonnmotion.go [hunk 3]
```diff
@@ -156,27 +164,43 @@ func (b Bonnmotion) Generate(exp experiment.Experiment) {
...
+	}
+	defer func() {
+		if cerr := b.stepFile.Close(); cerr != nil {
+			logger.Error("Error closing step file:", cerr)
+		}
+	}()
```
Comment (SA8R1NA) [inline internal/outputgenerators/bonnmotion.go:182]: do you need the ()?

File: internal/outputgenerators/bonnmotion.go [hunk 4]
```diff
@@ -156,27 +164,43 @@ func (b Bonnmotion) Generate(exp experiment.Experiment) {
...
 		logger.Error("Not allowed to write step file!")
 		return
 	}
+	b.stepFile, err = os.OpenFile(stepFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		logger.Error("Error opening step file:", err)
+		return
+	}
+	defer func() {
+		if cerr := b.stepFile.Close(); cerr != nil {
```
Comment (SA8R1NA) [inline internal/outputgenerators/bonnmotion.go:179]: if possible please define cerr outside of the if statement for readabillity reason

File: internal/parser/node_groups.go [hunk 5]
```diff
@@ -17,9 +17,28 @@ type inputNodeGroup struct {
 	NodesType     any `default:"PC"`
 	Networks      []string
 
-	MinSpeed any `default:"0"`
-	MaxSpeed any `default:"1"`
-	MaxPause any `default:"0"`
+	MinSpeed              any `default:"1"`
```
Comment (SA8R1NA) [inline internal/parser/node_groups.go:20]: is somewhere described what the values do?

File: internal/outputgenerators/bonnmotion.go (hunk 1 above)
Comment (ebinf) [reply internal/outputgenerators/bonnmotion.go:117]: This has technical reasons. movementModel is only of the correct type (which is switched over), if declared inline in the switch statement.

File: internal/outputgenerators/bonnmotion.go (hunk 2 above)
Comment (ebinf) [reply internal/outputgenerators/bonnmotion.go:176]: In fact, the code is consistent. `return nil` is used for functions that return an `error` as only return value, `return` without `nil`) is used for functions, that don't return anything

File: internal/outputgenerators/bonnmotion.go (hunk 3 above)
Comment (ebinf) [reply internal/outputgenerators/bonnmotion.go:182]: Yes, these are needed, as the function is declared inline and instantly called.

File: internal/outputgenerators/bonnmotion.go (hunk 4 above)
Comment (ebinf) [reply internal/outputgenerators/bonnmotion.go:179]: That would technically be possible, but would be inconsistent with [core.go](https://github.com/stg-tud/bp2022_netlab/blob/main/internal/outputgenerators/core.go#L257), [corremulab.go](https://github.com/stg-tud/bp2022_netlab/blob/main/internal/outputgenerators/coreemulab.go#L83) and [theone.go](https://github.com/stg-tud/bp2022_netlab/blob/main/internal/outputgenerators/theone.go#L158).

File: internal/parser/node_groups.go (hunk 5 above)
Comment (ebinf) [reply internal/parser/node_groups.go:20]: The meanings are described in the [wiki](https://github.com/stg-tud/bp2022_netlab/wiki/Node-Groups#randomwaypoint). But that's a good point, I forgot to remove the unused ones, now that I removed the additional movement patterns.

Comment (ebinf) [reply internal/parser/node_groups.go]: Should be fixed now

**Review by SA8R1NA — APPROVED** [review]

**Pull request merged**